### ISSUE TITLE
Support custom dialects

### ISSUE THREAD

**Issue by camsaul** [jkk/honeysql#401]
The default dialects cover most real-world use cases but there are situations where people might need to do something beyond these. Here are some motivating examples:

- H2 uppercases unquoted identifiers by default. For historic reasons related to Liquibase, Metabase H2 application DBs use all-uppercase identifiers. We have a custom dialect whose quote function combines the `:ansi` quote function with a call to `upper-case` to support this situation.
- As an application DB Metabase supports H2, MySQL/MariaDB, and Postgres; in tests we often have multiple app DBs in play at the same time (for example to test our loading data from an H2 database into a MySQL one). We have a custom `::application-db` quote function that looks at an `*application-db*` dynamic variable and chooses the right quote function based on that. In HoneySQL 2.x it would be nice to make `::application-db` our default dialect so we could use HoneySQL and have it quote things as expected based on the currently active application DB without explicitly specifying the dialect all over the place

We're currently hacking this stuff in with `alter-var-root`. A `register-dialect!` function would be nice if other people wanted to be able to do similar stuff for whatever reason they might need to

**Comment by seancorfield** [comment]
One of the issues to figure out here is that dialects in HoneySQL (2.x) are more than just a quoting function:

```clojure
(def ^:private dialects
  (reduce-kv (fn [m k v]
               (assoc m k (assoc v :dialect k)))
             {}
             {:ansi      {:quote #(strop \" % \")}
              :sqlserver {:quote #(strop \[ % \])}
              :mysql     {:quote #(strop \` % \`)
                          :clause-order-fn #(add-clause-before % :set :where)}
              :oracle    {:quote #(strop \" % \") :as false}}))
```

And this may change over time as new differences in behavior between dialects are discovered and need to be addressed.

For now, I could document that a dialect is a hash map with at least a `:quote` function and make the `strop` function public (it doubles-up the closing "quote" if it finds it inside the item being quoted), and also document the optional `:clause-order-fn`.

Then `dialects` would need to become an `atom` and I could add `register-dialect!` to `assoc` in the new hash map under the supplied dialect keyword (with some validation on its contents).

The downsides are:

* The current structure of dialects changes from an implementation detail to an API that should not be broken,
* If new differences are encountered between dialects that require additions to this structure, it could break any custom dialects that people build.

The latter might be partly mitigated by adding query functions that allow the current definition of a dialect to be obtained so that customizations based solely on wrapping functions -- as @camsaul describes -- could then be built and registered, even if the internals of a dialect change.

**Comment by seancorfield** [comment]
You can try this out on **develop** or in the latest 2.2.999-SNAPSHOT (be aware I'm switching to 2.3.x for the upcoming release so the next snapshot will be 2.3.999-SNAPSHOT).

It still needs tests and docs but you should be able to do something like:

```clojure
(sql/register-dialect! ::application-db (update (sql/get-dialect :ansi) :quote comp clojure.string/upper-case))
```

(although I guess you may want the `sql/upper-case` function exposed as public to avoid locale issues?)

**Comment by seancorfield** [comment]
Bug fixes, docs, and tests available on **develop** now or in the latest **2.3.999-SNAPSHOT**.
